### DISCUSSION TITLE
A question about the reference to the MaterialProperty in Material class

### DISCUSSION THREAD

**Discussion by TJT-post95** [idaholab/moose#21860] [Q&A General] [answered]
For example, a no-const reference _diffusivity of MaterialProperty<Real> ie declared in Example 8, and it is initialized by function declareProperty<Real>("diffusivity"). However, the parameter in the function is only a string for the name of property. So, does _diffusivity[_qp] have the assignment operations during initialization? Would it also work if we just declare MaterialProperty<Real> _diffusivity directly? Continuing to learn MOOSE, and looking forward to a reply.

Thank you!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Sorry for the delay we missed this. 
A Material object declares new material properties. Then this material also provides the definition to compute the vlaue of these properties. Computing and setting the value, done with `diffusivity[_qp] = ` is not a const-compatible operation, hence the on-const reference held by the object

Guillaume

**Reply by TJT-post95** [reply]
Thank you!
According to the definition of MaterialProperty class, the object of MaterialProperty class must be referenced to avoid the shallow copy problem. According to my current understanding, this may be the main reason that we don't declare `MaterialProperty _diffusivity;` directly?

**Reply by GiudGiud** [reply]
It s more that if you don’t take a reference, then you are making a copy of the property. 

If you make a copy of the property in your kernel for example, then when the Material object updates the properties, you don’t see the updated values in the copy

Guillaume

**Reply by TJT-post95** [reply]
I agree with your explanation, so we use constant reference in the Kernel. But in our own Material class, we declare the member variable `MaterialProperty<Real> & _diffusivity;` rather than `MaterialProperty<Real>  _diffusivity;`. After reading the source file of MaterialProperty class, now I think this may be a choice made by MOOSE to facilitate development and avoid the shallow copy problem?

**Reply by hugary1995** [reply]
It's not const because you want to write to it. It's a reference because because all material properties are registered and managed in a warehouse, thus the object declaring it only holds a reference. There is no shallow copy problem here.

**Reply by TJT-post95** [reply]
Got! Thank you Guillaume and hugary1995.
